Feature 579: Remove io-ts

daniel-va opened the pull request:
Resolves #579.

Refactors the application's models to no longer rely on the `fp-ts` and the `io-ts` libraries.

In the process of this refactoring, the models were consolidated. I renamed fields and restructured some relations to closer reflect their actual purpose. An overview of what has changed in regards to the DB mapping can be found in #613.

Due to the models being used in every part of the application, this refactoring concerns almost every part of the code. I have tested the application's behavior, though I can't ensure that I've gotten every bug yet (I heard automated tests would be great to not have to do this manually).

As a result of this PR, the two functional libraries remain in use at two places:
1. The edit view of geometries, which is due for refactoring at a future date. Not that as a side effect of the refactoring, I managed to fix the bug where saving was not possible on the geometry edit page.
2. The mapping of LV95 to WGS coordinates for the map is still backed by some functional types. This should be a relatively small refactoring, but due to the complexity of that mapping, I did not want to change it here.

Have fun :smile:

Tugark reviewed (changes requested):
**WIP - I will add more comments here after more testing (because I don't have DEV access right now), but I wanted to make sure my changes are not lost. IGNORE FOR NOW**

**In general,** I really like the changes and I'm glad we finally got rid of (most of) `io-ts`. I also like the additional code refactorings you did - however, because the `io-ts` part in itself is extremely large, I'd prefer to not have other unrelated style fixes as well. This made it quite hard to check what changed due to altered types or what changed because you stumbled upon ugly code. Also, I'm not sure if the rename from `study` to `location` or `geometry` is okay from a domain-perspective; I fully understand and support it, but if `study` is an actual domain term, I'm not sure we should rename all references in our code.

While testing, I found the following issues, so they somehow broke _somewhere_ in your changes (**note:** I will check them on DEV once I have access to see if they are new,  so please ignore for now).

**Editor**
* **Contacts:** I can no longer link a contact; entering a name in the modal does not trigger a search request anymore (as it does with references)
* **Contacts:** Changing the "Kontaktart" of an existing contact is not persisted.
* **Files:** Deleting files is no longer reflected in the GUI; they should be styled with a strikethrough
* **Files:** Export functionality does not work (i.e. nothing happens when you click on export)
* **Files:** There is a weirrd bug for assets with multiple files - Check File A -> Delete Button is enabled -> Uncheck File A -> Delete Button is disabled -> Check File B -> Delete Button can no longer be enabled, even if File A is checked
* **Files:** Adding a file is not visible until after save (it is added however)
* **Delete:** Deleting an asset shows the confirmation modal; the asset is not deleted however
* **Alternative IDs:** Altering an alternative ID (e.g. changing the title) does not trigger a dirty form, so it cannot be saved

**Asset View Sidebar**
* **National Interest:** If an asset has more than one national interest, and has been saved twice, all interests are displayed multiple times
![image](https://github.com/user-attachments/assets/8dd63e44-14e8-4bc8-993c-715582fce0c9)
* **National Interest:** Missing linebreaks
* **Languages:** If an asset has multiple languages, the display is off
![image](https://github.com/user-attachments/assets/47440661-b5d7-4224-98d1-5bb7d2b7a02f)

daniel-va commented on apps/client-asset-sg/src/app/i18n/de.ts:
```diff
@@ -106,7 +106,7 @@ export const deAppTranslations = {
     geometryCode: {
       Point: 'Punkt',
-      LineString: 'Linie',
+      Line: 'Linie',
```
I renamed it back to `LineString` now, which makes a lot more sense and is generally easier to use. I think I worked backward here from renaming `location`/`trace`/`area` and these names were caught in the fallout.

daniel-va commented on libs/client-shared/src/lib/pipes/reference.pipe.ts:
```diff
@@ -0,0 +1,71 @@
+import { Subscription } from 'rxjs';
+import { selectReferenceData } from '../state/app-shared-state.selectors';
+
+@Pipe({ name: 'reference', pure: false, standalone: true })
```
So, first off, this pipe is mainly used as replacement for [`namedValuePipe`](https://github.com/swisstopo/swissgeol-assets-suite/pull/619/files#diff-9cfc19a64671046f08209f817974b0d8ae58333fd28d3b077b84dd2fbf492be0), which was itself impure. This means `reference` shouldn't at the very least not degrade the application's performance any further.

Regardless, I see your point and would very much prefer using pure pipes here.
I see two ways we could make this pure / more performant.

1. Ensure that reference data is always loaded before rendering the rest of the application.
2. Move the reference data out of the API/DB (either into translations files or just make them static assets).

I think both of these are out-of-scope for this issue (especially 2).

The only other opportunity for caching/memoization I see here is to cache each value/key mapping locally. I'm not sure on the benefits of that, though. The pipe currently performs 4 key lookups (which should be O(1)). Memoization would reduce the best case to a single lookup, but add cache complexity and still be O(1).

Tugark replied:
Sorry, I did not catch that it's just a replacement for another impure pipe... There were too many other changes XD

How much effort would no 1 be? But I'm fine either way, but maybe you could add another ticket in the backlog and reference it in the cleanup ticket, so we don't forget this.

(there are multiple other occasions where we do this, so this in itself would be another refactor)

Tugark replied:
I see. 

So another "hack" solution would be to pass the current language to the pipe as an additional parameter, because then, the pipe could be pure, since the only time we need to recalculate the value is when the language parameter changes.

daniel-va commented on libs/shared/v2/src/lib/models/contact.ts:
```diff
@@ -10,15 +10,31 @@ export interface Contact extends Model<ContactId> {
 export type ContactData = Data<Contact>;
 
 export const AssetContactRoles = ['author', 'initiator', 'supplier'] as const;
-export type AssetContactRole = (typeof AssetContactRoles)[number];
+
+export enum AssetContactRole {
+  Author = 'author',
+  Initiator = 'initiator',
+  Supplier = 'supplier',
+}
```
I removed the `AssetContactRoles` array and replaced its usages by `Object.values(AssetContactRole)`.

Tugark commented on libs/client-shared/src/lib/utils/wgs.ts:
```diff
@@ -1,8 +1,15 @@
 
 import { WGS, WGSLat, WGSLng, isoWGSLat, isoWGSLng } from '../models';
 
+export const coordinateToWGS = (coordinate: Coordinate): WGS => {
```
Yes exactly. I just wanted to make sure that it still is correct because I stumbled over it and it still irks me XD

daniel-va replied:
Just so I understand this, Is this actually a bug/problem then? Or rather, what would be the solution? Should we adjust our code to flip x/y to align with Postgis/OL?

daniel-va commented:
@Tugark I've implemented your suggested changes and fixed the bugs you've listed. As for your other points:

- I agree that this refactoring is probably too large and could have been done in smaller parts - especially considering that someone has to review all these. I was kind of hard-focussed on doing a large, single-sweep of the entire model. Something for me to take away and do better next time.

- The rename from `study` to `geometries` should be very unproblematic. The term is not used by anyone at swissgeol (as far as I am aware). For example, the UI of this application always uses the term "Geometries", and never once uses "Study".

  To me, it seems "Study" is more confusing than anything. A large part of this issue was, to me, to make the code/models easier to understand for the developers. The Study model has long been one part of the application which caused quite a bit of confusion, and i hope this change solves some of that. 
  
As for testing, I would suggest to make more general tests for now, and then a big, application-encompassing test before we next deploy to INT. That is probably necesssary either way, seeing as how we are currently handling our bug/feature issues.

TIL-EBP commented on libs/shared/v2/src/lib/models/asset.ts:
```diff
@@ -1,130 +1,188 @@
-}
+// TODO There are only 2 assets with an entry in `type_nat_rel` and `is_nat_rel = FALSE`.
+//      I would assume that whenever a `type_nat_rel` exists, `is_nat_rel` should be true.
+//      This would enable use to remove the `is_nat_rel` field.
```
Should we discuss this with marcel?

daniel-va replied:
I opened #628 for it. I don't think we need to prioritize this all that much.

Tugark commented:
@daniel-va I tested it and most of it is now working properly. However, I noticed that you can no longer add multiple roles to a contact when editing it:
![image](https://github.com/user-attachments/assets/a56f6158-6860-4877-8dd7-6ec5d794a721)
It will always store only one of the roles. I _think_ on dev, this is working, but I cannot test it due to known reasons :)

1. Open an asset, move to contacts
2. Open existing contacts and add an additional role
3. Store it, and it will not add the selected roles

Also, the same happens when you add a new contact, it will only add one of the roles.

Either way, you could maybe fix this, since it's a rather ugly bug.

TIL-EBP reviewed (commented):
I didn't go into the deatil of the code as much as Lukas but rather focused on testing. 
Apart from what Lukas found, I found this:
- Geometries: There is no Geometry preselected when opening the Tab anymore

daniel-va commented on apps/server-asset-sg/src/features/assets/files/file.repo.ts:
```diff
@@ -48,29 +40,22 @@ export class FileRepo
-        type: data.type,
-        legalDocItemCode: data.legalDocItemCode,
+        type: data.legalDocCode === null ? 'Normal' : 'Legal',
+        legalDocItemCode: data.legalDocCode,
```
I added it to #613, but I don't want to touch the DB in this issue.

daniel-va commented on apps/server-asset-sg/src/features/assets/search/asset-search.service.ts:
```diff
@@ -222,28 +207,28 @@ export class AssetSearchService {
+        buckets: AggregationBucket<ContactId>[];
       };
-      assetKindItemCodes: {
+      kindCodes: {
```
My reasoning here is that you are in an asset, so its an asset's kind code. Its the same prefixing every ID column with the models name - if you think this is confusing I can add it back, I personally find this to be cleaner.

daniel-va commented on libs/shared/v2/src/lib/schemas/reference-data.schema.ts:
```diff
@@ -0,0 +1,44 @@
+
+  @IsArray()
+  @IsString({ each: true })
+  assetTopics!: LocalizedItem[];
```
`assetsFormat` was a typo, I renamed it to be singular, like anything else.

We never define a proper naming scheme here (mostly because these names, which are actually readable and understandable, are only here since this issue 😏). My reasoning here was I just prefix everyone of these kind-of-enums-but-not-really by the type that are used in. The only type that doesn't conform to this naming scheme is `LegalDocCodes`, which is a bit special either way.

I actually attempted to drop the `asset` prefix at first, but that got confusing because you `kinds` and `contactKinds`. However, As long as we use a standardized notation everyway, I'm fine either way.

daniel-va commented on libs/asset-viewer/src/lib/services/geometry.service.ts:
```diff
@@ -0,0 +1,55 @@
+        map((line) => {
+          const [id, assetId, geometryType, accessType, x, y] = line.split(';');
+          return {
+            id: `study_${id}` as GeometryId,
```
I mean, yes please, but since we are planning to go over geometries soon-ish anyways, I wouldn't do this here. Not like this PR isn't bloaded yet either.

daniel-va commented on libs/shared/v2/src/lib/utils/functions.ts:
```diff
@@ -0,0 +1,5 @@
+export const error = (message: string): never => {
```
I usually use something like this file for a couple simple functions, it is meant to be generic 😄 What name would you recommend using here?

Tugark commented on libs/asset-viewer/src/lib/components/asset-picker/asset-picker.component.ts:
```diff
@@ -53,24 +51,23 @@ const initialState: AssetPickerState = {
+  private readonly ngZone = inject(NgZone);
 
-  @ViewChild('pickerContainer') _pickerContainer!: ElementRef<HTMLElement>;
+  @ViewChild('pickerContainer') readonly _pickerContainer!: ElementRef<HTMLElement>;
```
That's what I assumed, and I'm happy that we're removing them now :D

daniel-va commented on libs/asset-editor/src/lib/components/asset-editor-tabs/asset-editor-legacy-data/asset-editor-legacy-data.component.ts:
```diff
@@ -8,5 +8,9 @@ import { AssetEditDetail } from '@asset-sg/shared';
-  @Input() asset: AssetEditDetail | null = null;
+  @Input() asset: Asset | null = null;
+
+  get legacyData(): AssetLegacyData | null {
+    return this.asset?.legacyData ?? null;
+  }
```
Fair enough - I've reverted the change.
